askama/src/filters/alloc.rs

```diff
@@ -376,39 +512,61 @@ pub fn trim<T: fmt::Display>(s: T) -> Result<String> {
 /// # }
 /// ```
 #[inline]
-pub fn indent(s: impl fmt::Display, width: usize) -> Result<String, fmt::Error> {
-    fn indent(args: fmt::Arguments<'_>, width: usize) -> Result<String, fmt::Error> {
-        let mut buffer = String::new();
-        let s = if width >= MAX_LEN {
-            buffer.write_fmt(args)?;
-            return Ok(buffer);
-        } else if let Some(s) = args.as_str() {
-            if s.len() >= MAX_LEN {
-                return Ok(s.into());
-            } else {
-                s
-            }
+pub fn indent<S: fmt::Display>(source: S, width: usize) -> Result<Indent<S>, Infallible> {
+    Ok(Indent { source, width })
+}
+
+pub struct Indent<S> {
+    source: S,
+    width: usize,
+}
+
+impl<S: fmt::Display> fmt::Display for Indent<S> {
+    fn fmt(&self, dest: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self { ref source, width } = *self;
+        if width >= MAX_LEN || width == 0 {
+            write!(dest, "{source}")
         } else {
-            buffer.write_fmt(args)?;
-            if buffer.len() >= MAX_LEN {
-                return Ok(buffer);
-            }
-            buffer.as_str()
-        };
+            let mut buffer;
+            flush_indent(dest, width, try_to_str!(source => buffer))
+        }
+    }
+}
 
-        let mut indented = String::new();
-        for (i, c) in s.char_indices() {
-            indented.push(c);
+impl<S: FastWritable> FastWritable for Indent<S> {
+    fn write_into<W: fmt::Write + ?Sized>(
+        &self,
+        dest: &mut W,
+        values: &dyn crate::Values,
+    ) -> crate::Result<()> {
+        let Self { ref source, width } = *self;
+        if width >= MAX_LEN || width == 0 {
+            source.write_into(dest, values)
+        } else {
+            let mut buffer = String::new();
+            source.write_into(&mut buffer, values)?;
+            Ok(flush_indent(dest, width, &buffer)?)
+        }
+    }
+}
 
-            if c == '\n' && i < s.len() - 1 {
-                for _ in 0..width {
-                    indented.push(' ');
-                }
+fn flush_indent(dest: &mut (impl fmt::Write + ?Sized), width: usize, s: &str) -> fmt::Result {
+    if s.len() >= MAX_LEN {
+        return dest.write_str(s);
+    }
+
+    let mut prefix = String::new();
+    for (idx, line) in s.split_inclusive('\n').enumerate() {
+        if idx > 0 {
+            // only allocate prefix if needed, i.e. not for single-line outputs
+            if prefix.is_empty() {
+                prefix = format!("{: >1$}", "", width);
             }
+            dest.write_str(&prefix)?;
         }
-        Ok(indented)
+        dest.write_str(line)?;
     }
-    indent(format_args!("{s}"), width)
+    Ok(())
 }
```

Comment:
`|indent` is entirely re-implement, please pay extra attention.

Reply:
Implementation looks fine to me, but it doesn't seem like we have tests for it so please add some. :)

Reply:
It has some in `filter_blocks` and `filter_blocks_ws`. Also, there is the doc test.